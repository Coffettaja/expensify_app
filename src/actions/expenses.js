import uuid from 'uuid'
import database from '../firebase/firebase'

//  The action generators for expenses

// Add expense
// Takes an object as an argument. The object will be used to set the expense properties. Default values for the argument object and default properties also set. An expense gets an ID value from uuid library function. 
// export const addExpense = (
//   {
//     description = '',
//     note = '',
//     amount = 0,
//     createdAt = 0
//   } = {}
// ) => ({
//   type: 'ADD_EXPENSE',
//   expense: {
//     id: uuid(),
//     description,
//     note,
//     amount,
//     createdAt,
//   }
// })

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

// starts the process of adding an expense to the store
export const startAddExpense = (expenseData = {}) => {
  // returning functions is OK because of redux-thunk middleware
  // it also gives them the dispatch function as a parameter
  return (dispatch) => {
    // this has basically the same result as default values in addExpense but maybe easier to read
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData

    // First push the data to the database, and if that works, dispatch the action to the redux store
    const expense = { description, note, amount, createdAt }
    // returning a promise to make promise chaining possible (mainly for testing purposes)
    return database.ref('expenses').push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }))
      })
  }
}

/**
 * Generates an action object that states what expense should be removed by ID
 * @return {Object} an action object for removing expenses from the state
 * @param {string} id
 */
export const removeExpense = (id) => ({
  type: 'REMOVE_EXPENSE',
  id
})

/**
 * Generates an action object with id of expense to be edited and updates.
 *
 * @param {string} id of the expense to be edited
 * @param {Object} updates - states the new property values
 * @param {string} [updates.description] - new description
 * @returns {Object} with the type: 'EDIT_EXPENSE'
 */
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})