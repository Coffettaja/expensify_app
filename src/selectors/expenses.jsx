import moment from 'moment'
import { create } from 'domain';

// Gets the visible expenses in sorted order. Takes two arguments: expenses, which is the complete array of expenses, and filters that are used to filter the expenses shown.
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt)
      // If startDate is set on the calendar, check if the startDate is earlier than the createdAt of the expense, by day.
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
      const textMatch = expense.description.toLowerCase().includes  (text.toLowerCase())

      return startDateMatch && endDateMatch && textMatch
    })
    // Sorts the expenses by either date or amount
    .sort((a, b) => {
      if (sortBy === 'date') {
        return b.createdAt - a.createdAt
      }
      if (sortBy === 'amount') {
        // When sorting by amount, the highest amount obviously comes first.
        // However, if the amounts are equal, the newer expense comes first.
        if (b.amount - a.amount !== 0) {
          return b.amount - a.amount
        }
        return b.createdAt - a.createdAt
      }
    })
}

export default getVisibleExpenses