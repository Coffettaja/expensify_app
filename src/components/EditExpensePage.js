import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, startRemoveExpense } from '../actions/expenses'
import { SSL_OP_CRYPTOPRO_TLSEXT_BUG } from 'constants';

export class EditExpensePage extends React.Component {
  onSubmit = (newExpenseValues) => {
    this.props.editExpense(this.props.expense.id, newExpenseValues)
    this.props.history.push('/')
  }

  onRemoveClick = () => {
    this.props.startRemoveExpense(this.props.expense.id)
    this.props.history.push('/')
  }

  render() {
    if (!this.props.expense) {
      return (
        <div>
          <p>
            Looks like this expense does not exist :(
        </p>
          <Link to="/">Back to dashboard</Link>
        </div>
      )
    }
    return (
      <div>
        <ExpenseForm
          onSubmit={this.onSubmit}
          expense={this.props.expense}
        ></ExpenseForm>
        <button onClick={this.onRemoveClick}>Remove!</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(x => x.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    editExpense: (id, newExpenseValues) => dispatch(editExpense(id, newExpenseValues)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)