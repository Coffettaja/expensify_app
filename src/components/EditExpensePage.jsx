import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm.jsx'
import { editExpense, removeExpense } from '../actions/expenses.jsx'
import { SSL_OP_CRYPTOPRO_TLSEXT_BUG } from 'constants';

const EditExpensePage = (props) => (
  <div>
    <ExpenseForm 
      onSubmit={(expense) => {
        props.dispatch(editExpense(props.expense.id, expense))
        // console.log(expense)
        props.history.push('/')
      }} 
      expense={props.expense}
    ></ExpenseForm>
    <button onClick={(e) => {
      props.dispatch(removeExpense(props.expense.id))
      props.history.push('/')
    }}>Remove!</button>
  </div>
)

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(x => x.id === props.match.params.id)
  }
}

export default connect(mapStateToProps)(EditExpensePage)