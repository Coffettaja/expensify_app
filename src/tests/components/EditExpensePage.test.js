import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/EditExpensePage.jsx'
import expenses from '../fixtures/expenses'

let editExpense, removeExpense, history, expenseToEdit, wrapper

beforeEach(() => {
  editExpense = jest.fn()
  removeExpense = jest.fn()
  history = { push: jest.fn() }
  expenseToEdit = expenses[0]
  wrapper = shallow(<EditExpensePage editExpense={editExpense} removeExpense={removeExpense} history={history} expense={expenseToEdit}></EditExpensePage>)
})

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle editExpense on submit', () => {
  // the values don't actually matter in this case ...
  const newExpenseValues = { description: 'test description', amount: '100000' }
  wrapper.find('ExpenseForm').prop('onSubmit')(newExpenseValues)
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(editExpense).toHaveBeenLastCalledWith(expenseToEdit.id, newExpenseValues)
})

test('should handle removeExpense on button click', () => {
  wrapper.find('button').at(0).simulate('click') //prop('onClick')()
  expect(removeExpense).toHaveBeenLastCalledWith(expenseToEdit.id)
})