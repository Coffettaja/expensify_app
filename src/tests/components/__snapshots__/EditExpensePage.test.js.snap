// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 6000,
        "createdAt": 0,
        "description": "car",
        "id": "0",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove!
  </button>
</div>
`;
